### PR TITLE
More flexibility on numerical/string doc id and quantization

### PR DESCRIPTION
**Pull request by amallia**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -778,7 +783,9 @@ fn pisa_to_ciff_from_paths(
 #[derive(Debug, Deserialize)]
 struct JsonDoc {
     /// Collection docid. CIFF will automatically assign an integer ID to this.
+    /// Can be provided as either string or integer, will be converted to string.
     #[serde(default)]
+    #[serde(deserialize_with = "deserialize_id_to_string")]
```
Comment (elshize) [inline src/lib.rs:788]: Can we add a test for deserializing `JsonDoc` both with int and string?

File: src/lib.rs [hunk 2]
```diff
@@ -831,13 +866,51 @@ impl JsonlToCiff {
             File::open(input_path).with_context(|| format!("Cannot open {input_path:?}"))?;
         let total_input_size = input_file.metadata()?.len();
 
+        // If quantization is enabled, we need to find min/max first
```
Comment (elshize) [inline src/lib.rs:869]: Can we extract this to a function? Since we're reopening the file anyway, I would pass the _path_ into this function and open it inside this new function.

Once extracted, it should be very easy to write some simple unit tests for this. I can see three obvious test cases:
1. No documents - failure
2. Documents with only empty vectors - failure
3. Only vectors with scores <=0 - failure
4. Some scores >0 -- ok!

Example inputs:

1. Empty string

2.

```json
{"id": 0, "vector": {}}
{"id": 1, "vector": {}}
```

3.

```json
{"id": 0, "vector": {"a": 0.0, "b": -1.0}}
{"id": 1, "vector": {"c": -0.001, "d": -777}}
```

4.

```json
{"id": 0, "vector": {"a": 0.0, "b": 1.0}}
{"id": 1, "vector": {"c": 0.001, "d": -777}}
```

---

BONUS QUESTION: What happens when *only one* value is >0? Is this correct or should we fail?

File: src/lib.rs [hunk 3]
```diff
@@ -873,7 +946,24 @@ impl JsonlToCiff {
             // Sum of tf's in this doc => doc_length
             let mut doc_length = 0i64;
             for (term, score) in jdoc.vector {
-                let tf = score.round() as i32;
+                let tf = if self.quantize {
+                    // 8-bit scalar quantization: map [min_score, max_score] to [1, 256]
```
Comment (elshize) [inline src/lib.rs:950]: This also should be a separate function. And again, it will allow us to easily test different combinations and corner cases.

Interesting things to test:
* If score=min, then quantized score is 1
* If score=max, then quantized score is 255
* Some other value

File: src/lib.rs [hunk 4]
```diff
@@ -873,7 +946,24 @@ impl JsonlToCiff {
...
+                    // We use 1-256 to avoid zero values which get filtered out
```
Comment (elshize) [inline src/lib.rs:951]: Here, as well as on the line above, you say 256 but the actual max score is 255.

File: .github/workflows/rust.yml [hunk 5]
```diff
@@ -65,30 +65,6 @@ jobs:
       - run: cargo rustdoc --lib -- -D warnings
       - run: cargo rustdoc --bin pisa2ciff -- -D warnings
       - run: cargo rustdoc --bin ciff2pisa -- -D warnings
-  code-coverage:
```
[the remark this answers is not in the document]
Comment (amallia) [reply .github/workflows/rust.yml:68]: It's broken and also not very useful so I thought to remove it.
Anyway, if we want to fix it we probably need to rewrite it at this point.

[the remark this answers is not in the document]
Comment (elshize) [reply .github/workflows/rust.yml:68]: Sounds good.

File: src/lib.rs [hunk 6]
```diff
@@ -845,6 +851,99 @@ impl JsonlToCiff {
...
+            "max_score ({}) must be >= min_score ({})",
+            max_score,
+            min_score
+        );
+
+        if score <= 0.0 {
+            0 // Will be filtered out
+        } else if min_score == max_score {
+            // If all scores are the same, map to max value
+            MAX_QUANTIZED_VALUE
```
Comment (elshize) [inline src/lib.rs:891]: Just curious: would there be any difference between choosing max and min? If not, then wouldn't it be better to chose min? If yes, could you explain the difference?

File: src/lib.rs [hunk 7]
```diff
@@ -845,6 +851,99 @@ impl JsonlToCiff {
...
+    ///
+    /// # Returns
+    /// * `0` if score <= 0 (will be filtered out)
+    /// * `MAX_QUANTIZED_VALUE` if min_score == max_score (all scores identical)
+    /// * Quantized value in range [MIN_QUANTIZED_VALUE, MAX_QUANTIZED_VALUE] otherwise
+    ///
+    /// # Panics
+    /// Panics if min_score <= 0 or max_score <= 0 or max_score < min_score
+    fn quantize_score(score: f64, min_score: f64, max_score: f64) -> i32 {
+        assert!(
```
Comment (elshize) [inline src/lib.rs:870]: Generally, I feel pretty strongly about not panicking in a function like that. It's sort of ok here because it's a private function and an error really means a bug, so we would want to crash anyway (we are checking beforehand that the values should be correct). That said, I think this kind of design leads to less readable code that is harder to reason about.

I'm not necessarily insisting on a change here an now, but let me state my case anyway.

---

It is true that it is _typically_ recommended to panic when you discover an invalid state in your program (see more in [the book](https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html#to-panic-or-not-to-panic)). But I think we need to be more subtle than that. I think the _context_ of the panic is also important. While, if looking at the code as a whole, calling this function with `min_score < 0`, the panic is far from the source of the issue. This is a hint that perhaps it can be designed in a way that makes those check not necessary each time we quantize. Another hint is that we assert for each call, but the checked values do not change across the calls. So maybe a different approach is better.

A general thought about panics in code: yes, do panic if there's invalid state, but before you do, think if redesigning could eliminate the panic, decrease number of panics, or move the panic to a place where it makes more sense.

---

In this example, I would argue that min/max need to be checked once. How do we achieve this?

One way I can think of: instead of passing min/max to a function, we can have `ScoreQuantizer` struct with `quantize(score: f64)` method instead.


```rust
struct ScoreQuantizer {
    min: f64,
    max: f64,
}

impl ScoreQuantizer {
    fn new(min: f64, max: f64) -> ResultQuantizer {
        // asserts
        ScoreQuantizer{min, max}
    }

    fn quantize(&self, score: f64) -> i32 {
        // we already know min/max are ok, proceed with quantization logic
    }
}
```

Then:

```rust
let quantizer = if self.quantize {
    Some(Self::find_score_range(input_path).map(ScoreQuantizer::new)?)
} else {
    None
};
```
And finally:

```rust
let tf = match &quantizer {
    // 8-bit scalar quantization: map [min_score, max_score] to [1, 255]
    // We use 1-255 to avoid zero values which get filtered out
    Some(quantizer) => quantizer.quantize(score),
    // Assume scores are already pre-quantized integers
    None => score as i32,
};
```

File: src/lib.rs (hunk 6 above)
Comment (amallia) [reply src/lib.rs:891]: I think if `min == max ` then it does not make any difference. Sure, let's pick min as default

File: src/lib.rs (hunk 7 above)
Comment (amallia) [reply src/lib.rs]: you are amazing!

File: src/lib.rs (hunk 6 above)
Comment (elshize) [reply src/lib.rs]: Ok, thanks. Yeah, just to clarify, I suggested min because it's more space efficient in the end, though this is such an unexpected corner case, it shouldn't make any difference really...

**Review by elshize — APPROVED** [review]
Looks good! :fire:

**Comment by amallia** [comment]
Thanks Michal for the review!
